File: 0138-Copy_List_with_Random_Pointer/Copy_List_with_Random_Pointer.java
import java.util.*;
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) return null;
        HashMap<Node, Node> mapping = new HashMap<>();
        
        Node newHead = new Node(head.val);
        Node newTemp = newHead;
        Node temp = head;
        mapping.put(head, newHead);
        
        while(temp.next != null){
            newTemp.next = new Node(temp.next.val);
            
            temp = temp.next;
            newTemp = newTemp.next;
            mapping.put(temp, newTemp);
        }
        
        temp = head;
        newTemp = newHead;
        while(temp != null){
            if(temp.random != null){
                newTemp.random = mapping.get(temp.random);
            }
            
            newTemp = newTemp.next;
            temp = temp.next;
        }
        
        return newHead;
    }
}

File: 0138-CopyListWithRandomPointer/CopyListWithRandomPointer.go
package copylistwithrandompointer

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return head
	}

	mapping := make(map[*Node]*Node)

	newHead := &Node{Val: head.Val}
	newTemp := newHead
	temp := head

	mapping[head] = newHead

	for temp.Next != nil {
		temp = temp.Next
		newTemp.Next = &Node{Val: temp.Val}
		newTemp = newTemp.Next
		mapping[temp] = newTemp
	}

	temp, newTemp = head, newHead
	for temp != nil {
		if temp.Random != nil {
			newTemp.Random = mapping[temp.Random]
		}
		temp, newTemp = temp.Next, newTemp.Next
	}

	return newHead

}
